CVAT-UI: measure coverage.

dvkruchinin opened the pull request:
<!---
Copyright (C) 2020 Intel Corporation

SPDX-License-Identifier: MIT
-->

<!-- Raised an issue to propose your change (https://github.com/opencv/cvat/issues).
It will help avoiding duplication of efforts from multiple independent contributors.
Discuss your ideas with maintainers to be sure that changes will be approved and merged.
Read the [CONTRIBUTION](https://github.com/opencv/cvat/blob/develop/CONTRIBUTING.md)
guide. -->
CVAT-UI: measure coverage.

### Motivation and context
Measuring code coverage. Analysis and writing of texts on the basis of the analysis.

### How has this been tested?
<!-- Please describe in detail how you tested your changes.
Include details of your testing environment, and the tests you ran to
see how your change affects other areas of the code, etc. -->

### Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply.
If an item isn't applicable by a reason then ~~explicitly strikethrough~~ the whole
line. If you don't do that github will show incorrect process for the pull request.
If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I submit my changes into the `develop` branch
- [ ] I have added description of my changes into [CHANGELOG](https://github.com/opencv/cvat/blob/develop/CHANGELOG.md) file
- [ ] I have updated the [documentation](
  https://github.com/opencv/cvat/blob/develop/README.md#documentation) accordingly
- [ ] I have added tests to cover my changes
- [ ] I have linked related issues ([read github docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [ ] I have increased versions of npm packages if it is necessary ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
[cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning), [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [x] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2020 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

bsekachev commented:
@dvkruchinin 

CI is failed because of timeout.
Tests are performed too long in Firefox (only 15 of 33 are done).
Tests in Chrome haven't started.
Can it be related with coverage?

dvkruchinin commented:
> @dvkruchinin
> 
> CI is failed because of timeout.
> Tests are performed too long in Firefox (only 15 of 33 are done).
> Tests in Chrome haven't started.
> Can it be related with coverage?
> 
> Update:
> 
> I've just seen that you disabled tests in Chrome. Ok. It's explain why they haven't started.

Tests for Firefox took an unexpectedly long time to run after the code was instrumented.
For example test auth_page.js (the code is already instrumented):
Firefox: duration 1 minute, 59 seconds 
Chrome: duration 7 seconds  
This is why the timeout was exceeded.
I will try to find out the reason for this behavior.

dvkruchinin commented:
The PR has been redesigned to prepare for the implementation of coverage measurement. Running tests via Firefox is also temporarily disabled.

bsekachev commented on package.json:
```diff
@@ -30,10 +30,22 @@
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "test": "echo \"Error: no test specified\" && exit 1",
+    "coverage": "npm run instrument && npm run cp && npm run rm",
+    "instrument": "nyc instrument cvat-ui cvat-ui_cov && nyc instrument cvat-canvas cvat-canvas_cov && nyc instrument cvat-data cvat-data_cov && nyc instrument cvat-core cvat-core_cov",
```
The config is based on ``@istanbuljs/nyc-config-typescript``, but cvat-core and cvat-data are currently JS written. Isn't it a problem?

dvkruchinin replied:
I remove this line from .nycrc. And add to package.json @istanbuljs/nyc-config-babel. 
As we discussed earlier, this has had a positive effect on file instrumentation. In particular, a larger number of .tsx files were included in the instrumentation.

bsekachev commented on .nycrc:
```diff
@@ -0,0 +1,19 @@
+        ".tsx"
+    ],
+    "exclude": [
+        "**/*.config.*"
```
What have you excluded?
BTW I suspect that node_modules could also be instrumented and this might be a reason of performance degradation. What do you think?
And maybe disable instrumentation of ``tests/**/.js``?

dvkruchinin replied:
node_modules not included in the instrumentation. This is the default setting for nyc.
Added the directories to the exclude: 
3rdparty
tests

bsekachev commented:
In the branch ``bs/test_travis`` (forked from this branch) I updated ``.travis.yml`` file. There I've added if condition that runs different code based on what triggered the built (push, pr or schedule [cron]). So, on push, pr it runs only cypress tests in Chrome, on schedule it runs also tests in Firefox and tests in Chrome after code instrumenting. Also I've added regular daily built on travis-ci.org. 

In test PR #2257 it looks to work as expected. So, do you think we can use this solution to built optimization?

dvkruchinin commented:
> In the branch `bs/test_travis` (forked from this branch) I updated `.travis.yml` file. There I've added if condition that runs different code based on what triggered the built (push, pr or schedule [cron]). So, on push, pr it runs only cypress tests in Chrome, on schedule it runs also tests in Firefox and tests in Chrome after code instrumenting. Also I've added regular daily built on travis-ci.org.
> 
> In test PR #2257 it looks to work as expected. So, do you think we can use this solution to built optimization?

It looks great. I think this is quite a working option for optimization.

bsekachev commented:
@dvkruchinin 

In this case,  could you to update ``.travis.yml`` in this PR? (I will close another one)

dvkruchinin commented:
> @dvkruchinin
> 
> In this case, could you update `.travis.yml` in this PR? (I will close another one)

Sure

bsekachev commented:
Great! Let's look how it will work for a couple of days and decide whether we need something else here or not.
